File: queue.cpp
#include "queue.h"
#include <iostream>
#include <fstream>

using namespace std;

Queue::Queue() : head(nullptr), tail(nullptr) {}

Queue::~Queue() {
    while (head) {
        QNode* temp = head;
        head = head->next;
        delete temp;
    }
}

void Queue::push(string val) {
    QNode* newNode = new QNode{val, nullptr};
    if (!head) {
        head = tail = newNode;
    } else {
        tail->next = newNode;
        tail = newNode;
    }
}

string Queue::pop() {
    if (!head) return "[QUEUE_EMPTY]";
    QNode* temp = head;
    string val = head->value;
    head = head->next;
    if (!head) tail = nullptr;
    delete temp;
    return val;
}

string Queue::peek() const {
    if (!head) return "[QUEUE_EMPTY]";
    return head->value;
}

void Queue::print() const {
    if (!head) {
        cout << "Очередь пуста." << endl;
        return;
    }
    cout << "Очередь: ";
    QNode* curr = head;
    while (curr) {
        cout << curr->value;
        if (curr->next) cout << " -> ";
        curr = curr->next;
    }
    cout << endl;
}

bool Queue::isEmpty() const {
    return head == nullptr;
}

File: ds/queue.go
package ds

import "fmt"

type QNode struct {
	Data string
	Next *QNode
}

type Queue struct {
	Front *QNode
	Rear  *QNode
}

func NewQueue() *Queue {
	return &Queue{}
}

func (q *Queue) IsEmpty() bool {
	return q.Front == nil
}

func (q *Queue) Push(val string) {
	newNode := &QNode{Data: val, Next: nil}
	if q.IsEmpty() {
		q.Front = newNode
		q.Rear = newNode
	} else {
		q.Rear.Next = newNode
		q.Rear = newNode
	}
}

func (q *Queue) Pop() string {
	if q.IsEmpty() {
		return "[QUEUE_EMPTY]"
	}
	val := q.Front.Data
	q.Front = q.Front.Next
	if q.Front == nil {
		q.Rear = nil
	}
	return val
}

func (q *Queue) Peek() string {
	if q.IsEmpty() {
		return "[QUEUE_EMPTY]"
	}
	return q.Front.Data
}

func (q *Queue) Print() {
	if q.IsEmpty() {
		fmt.Println("Очередь пуста.")
		return
	}
	fmt.Print("Очередь (начало -> конец): ")
	for cur := q.Front; cur != nil; cur = cur.Next {
		fmt.Printf("\"%s\" ", cur.Data)
	}
	fmt.Println()
}